# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

导入数据分析所需要的库，平通过pandas的`read_csv`函数，将原数据文件"e_commerce.csv"里的数据内容，解析为DataFrame，并赋值给变量`df`

In [1]:
import pandas as pd
df = pd.read_csv("e_commerce.csv")
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

#### **评估数据整洁度**

In [2]:
df.sample(10)

InvoiceNo StockCode                          Description  Quantity  \
342364    566860     22548         HEADS AND TAILS SPORTING FUN        12   
296265    562880     22646   CERAMIC STRAWBERRY CAKE MONEY BANK        72   
74542     542433     21829           DINOSAUR KEYRINGS ASSORTED        36   
62098     541496     37340        MULTICOLOUR SPRING FLOWER MUG         1   
71803     542215     22179              SET 10 LIGHTS NIGHT OWL         2   
82116     543188     21716      BOYS VINTAGE TIN SEASIDE BUCKET         8   
235160    557626    84970L     SINGLE HEART ZINC T-LIGHT HOLDER        12   
88954     543826     21898             POTTING SHED ROSE CANDLE         4   
265933    560234     22918                  HERB MARKER PARSLEY        12   
80694     543054     22967  SET 3 SONG BIRD PAPER EGGS ASSORTED         6   

            InvoiceDate  UnitPrice  CustomerID         Country  
342364  9/15/2011 12:40       1.25     17002.0  United Kingdom  
296265  8/10/2011 10:53       1.25     15838.0  United Kingdom  
74542   1/28/2011 10:45       0.21     12726.0          France  
62098   1/18/2011 15:00       8.29         NaN  United Kingdom  
71803   1/26/2011 12:27       6.75     12472.0         Germany  
82116    2/4/2011 11:45       2.55     12567.0          France  
235160  6/21/2011 14:29       0.95     14175.0  United Kingdom  
88954    2/14/2011 9:41       2.46         NaN  United Kingdom  
265933  7/17/2011 11:21       0.65     16180.0  United Kingdom  
80694    2/3/2011 10:39       2.95     13394.0  United Kingdom

结论：抽样10行结果来看，符合“每列是一个变量，每行是一个观察值，每个单元格十一个值”，具体看来每行是关于某商品的一次交易，每列是交易相关变量，因此不存在结构性问题

#### **评估数据干净程度**

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


结论：通过输出结果看来，数据总共有541909条观察值，而 ``Description（540455）, CustomerID（406829）`` 这两个变量存在缺失值。  
     另外，``InvoiceDate`` 变量的数据类型应该为日期，而`CustomerID`变量无需统计运算，数据类型应该为字符串，应该进行数据格式转换。

#### **评估缺失数据**

根据条件提取出`Description`列中的缺失值

In [4]:
df[df['Description'].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条交易数据缺失Description变量值。  
从输出结果看来，缺失`Description`的交易数据，`UnitPrice`都为0。为了验证猜想我们增加筛选条件，看是否存在`Description`变量缺失且`UnitPrice`不为0的数据

In [5]:
df[(df['Description'].isnull()) & (df['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来结果为0条，说明假设成立，缺失`Description`值的数据，同样也不具备有效的`UnitPrice`值。
`Description`与``UnitPrice``都是商品分析的重要变量，如果他们同时缺失/无效，我们认为这些数据无法提供有效意义，因此这些数据可以在后续进行删除。

根据条件提取出`CustomerID`列中的缺失值

In [6]:
df[df['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

`CustomerID`表示客户编号，不受分析畅销产品的必要变量。并且从结果来看，有些`CustomerID`缺失的交易数据仍然有小，因此保留此变量为空的观察值。

#### **评估重复数据**

根据数据变量含义看来，虽然`InvoiceNO`,`StockCode`,`CustomerID`都是唯一标识符但是：  
一次交易可能包含多件商品，因此`InvoiceNO`可以存在重复  
不同交易可以包含同件商品，因此`StockCode`可以存在重复  
顾客可以进行多次交易或下单多个商品，因此`CustomerID`可以存在重复

#### **评估不一致数据**

不一致数据可能存在于`Country`变量中，我们可以查看是否存在多个不同值知道同一国家的情况

In [7]:
df['Country'].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


从筛选结果看来，`USA`,`United States`均在表示美国，`United Kingdom`,`UK`,`U.K.`均在表示英国，因此需要在数据清洗阶段对这些值进行统一，只保留一个指代值。

#### **评估无效或错误数据**

通过`describe`方法，对数值统计信息快速了解

In [8]:
df.describe(include = 'all')

InvoiceNo StockCode                         Description       Quantity  \
count     541909    541909                              540455  541909.000000   
unique     25900      4070                                4223            NaN   
top       573585    85123A  WHITE HANGING HEART T-LIGHT HOLDER            NaN   
freq        1114      2313                                2369            NaN   
mean         NaN       NaN                                 NaN       9.552250   
std          NaN       NaN                                 NaN     218.081158   
min          NaN       NaN                                 NaN  -80995.000000   
25%          NaN       NaN                                 NaN       1.000000   
50%          NaN       NaN                                 NaN       3.000000   
75%          NaN       NaN                                 NaN      10.000000   
max          NaN       NaN                                 NaN   80995.000000   

             InvoiceDate      UnitPrice     CustomerID         Country  
count             541909  541909.000000  406829.000000          541909  
unique             23260            NaN            NaN              41  
top     10/31/2011 14:41            NaN            NaN  United Kingdom  
freq                1114            NaN            NaN          495266  
mean                 NaN       4.611114   15287.690570             NaN  
std                  NaN      96.759853    1713.600303             NaN  
min                  NaN  -11062.060000   12346.000000             NaN  
25%                  NaN       1.250000   13953.000000             NaN  
50%                  NaN       2.080000   15152.000000             NaN  
75%                  NaN       4.130000   16791.000000             NaN  
max                  NaN   38970.000000   18287.000000             NaN

In [9]:
df[df['Quantity'] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

从筛选结果来看，`Quantity`变量为复数的观察值，`InvoiceNo`似乎均以C开头，表示订单被取消。  
但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在`Quantity`变量为负数且`InvoiceNo`不已C开头的观察值

In [10]:
df[(df['Quantity'] < 0) & (df['InvoiceNo'].str[0] != 'C')]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被证实错误，因为还存在1336行数据`Quantity`变量为复数且`InvoiceNo`不以C开头  
但根据以上输出结果，这些筛选出的观察值的`UnitPrice`观察值均为0，因此增加`UnitPrice`的条件验证

In [11]:
df[(df['Quantity'] < 0) & (df['InvoiceNo'].str[0] != 'C') & (df['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当Quantity 变量为负数的时候，观蔡值满足以下条件之一  
- `Invoiceso` 以C开头，表示订单被取消  
- `Uniterice` 为0，说明单价为O英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范四内，因此我们将在数据清理步骤，将`Quantity`为负数的观察值删除。

接下来筛选`UnitPrice`为复数的观察值。

In [12]:
df[df['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

从输出结果来看，`UnitPrice`为负数的观察值都是坏账调账，不属于实际商品交易数据，因此也在数据清洗阶段将其删除。

## 清理数据

根据前面评估部分得到的结论，清洗阶段步骤包括：  
- 把`InvoiceDate`变量数据类型转换为日期格式
- 把`CustomerID`变量数据类型转化为字符串
- 把`Description`变量缺失的观察值删除
- 把`Country`变量值`USA`转化为`United States`, `UK`,`U.K.`转化为`United Kingdom`
- 把`Quantity`变量值为负数的观察值删除
- 把`UnitPrice`变量值为负数的观察值删除

为区分开经过清理的数据和原始数据，我们创建新的变量`cleaned_data`，让它成为`df`的副本。之后的清洗步骤都被使用在`cleaned_data`上

In [13]:
cleaned_data = df.copy()

把`InvoiceDate`变量数据类型转换为日期格式

In [14]:
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'])
cleaned_data['InvoiceDate']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把`CustomerID`变量数据类型转化为字符串

In [15]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].astype(str)
cleaned_data['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把`CustomerID`变量值结尾的".0"删除

In [16]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].str.slice(0,-2)
cleaned_data['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把`Description`变量缺失的观察值删除

In [17]:
cleaned_data.dropna(subset = ["Description"], axis = 0, inplace = True)

把`Country`变量值`USA`转化为`United States`, `UK`,`U.K.`转化为`United Kingdom`

In [18]:
cleaned_data['Country'].replace({'USA' : 'United States'}, inplace = True)
cleaned_data['Country'].replace(['UK','U.K.'], 'United Kingdom', inplace = True)

C:\Users\hp\AppData\Local\Temp\ipykernel_27468\4152214705.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  cleaned_data['Country'].replace({'USA' : 'United States'}, inplace = True)


把`Quantity`变量值为负数的观察值删除

In [19]:
cleaned_data = cleaned_data[cleaned_data['Quantity'] >= 0]

把`UnitPrice`变量值为负数的观察值删除

In [20]:
cleaned_data = cleaned_data[cleaned_data['UnitPrice'] >= 0]

观察清洗后数据并且采样30条

In [21]:
cleaned_data.sample(30)

InvoiceNo StockCode                          Description  Quantity  \
50419     540551     22186                 RED STAR CARD HOLDER         1   
35572     539436     84375        SET OF 20 KIDS COOKIE CUTTERS         1   
49659     540525     40016        CHINESE DRAGON PAPER LANTERNS        24   
39173     539631     22276        WASH BAG VINTAGE ROSE PAISLEY         1   
341314    566750     23317                BLUE REFECTORY CLOCK          2   
20266     538035     21929       JUMBO BAG PINK VINTAGE PAISLEY         1   
326849    565596     22086      PAPER CHAIN KIT 50'S CHRISTMAS         14   
282930    561671     22536           MAGIC DRAWING SLATE PURDEY       192   
475693    576910     22191               IVORY DINER WALL CLOCK        12   
316751    564756     21172                    PARTY METAL SIGN          3   
345164    567148     22663          JUMBO BAG DOLLY GIRL DESIGN        20   
82931     543275    90184B  AMETHYST CHUNKY BEAD BRACELET W STR         1   
5747      536874     21080  SET/20 RED RETROSPOT PAPER NAPKINS          1   
87583     543660     21001                 ROSE DU SUD WASHBAG          1   
472471    576688     22457      NATURAL SLATE HEART CHALKBOARD          6   
283959    561820     21672   WHITE SPOT RED CERAMIC DRAWER KNOB         1   
17254     537692     22909  SET OF 20 VINTAGE CHRISTMAS NAPKINS         4   
311717    564315    84997C      CHILDRENS CUTLERY POLKADOT BLUE         1   
124816    546992     22629                  SPACEBOY LUNCH BOX        128   
382048    569897     22139     RETROSPOT TEA SET CERAMIC 11 PC          2   
217839    555938     22621           TRADITIONAL KNITTING NANCY        12   
301860    563355     23229            VINTAGE DONKEY TAIL GAME          6   
366380    568721     22992               REVOLVER WOODEN RULER          1   
397981    571216     22645  CERAMIC HEART FAIRY CAKE MONEY BANK         2   
526488    580671     21564       PINK  HEART SHAPE LOVE BUCKET         24   
387530    570332     23322          LARGE WHITE HEART OF WICKER        12   
387377    570287     22621           TRADITIONAL KNITTING NANCY        12   
413508    572314     17003                  BROCADE RING PURSE         48   
15627     537638    90003B   ROSE COLOUR PAIR HEART HAIR SLIDES         1   
210610    555284     23282      FOLDING BUTTERFLY MIRROR IVORY         36   

               InvoiceDate  UnitPrice CustomerID         Country  
50419  2011-01-10 09:43:00       5.91          n  United Kingdom  
35572  2010-12-17 14:49:00       4.21          n  United Kingdom  
49659  2011-01-09 13:02:00       0.42      16833  United Kingdom  
39173  2010-12-20 15:03:00       5.91          n  United Kingdom  
341314 2011-09-14 15:41:00       9.95      15380  United Kingdom  
20266  2010-12-09 13:03:00       1.95      16065  United Kingdom  
326849 2011-09-05 14:27:00       2.95      16444  United Kingdom  
282930 2011-07-28 17:17:00       0.27      12830  United Kingdom  
475693 2011-11-17 09:51:00       7.65      12477         Germany  
316751 2011-08-30 10:33:00       4.13          n  United Kingdom  
345164 2011-09-16 15:23:00       2.08      18225  United Kingdom  
82931  2011-02-06 12:49:00       8.95      13584  United Kingdom  
5747   2010-12-03 11:35:00       0.85      16891  United Kingdom  
87583  2011-02-11 10:40:00       4.13          n  United Kingdom  
472471 2011-11-16 12:16:00       2.95      16717  United Kingdom  
283959 2011-07-29 16:00:00       3.29          n  United Kingdom  
17254  2010-12-08 10:40:00       0.85      14796  United Kingdom  
311717 2011-08-24 13:07:00       4.15      14730  United Kingdom  
124816 2011-03-18 13:18:00       1.65      16684  United Kingdom  
382048 2011-10-06 16:01:00       4.95      17813  United Kingdom  
217839 2011-06-08 09:58:00       1.65      14257  United Kingdom  
301860 2011-08-15 14:39:00       3.75      14156            EIRE  
366380 2011-09-28 16:24:00       4.13          n  United Kingdom  
397981 2011

## 保存清理后的数据

完成数据清洗，把干净数据保存在新的文件里，文件名为`'e_commerce_cleaned.csv'`

In [22]:
cleaned_data.to_csv('e_commerce_cleaned.csv', index = False)